File: edge/src/rules.mjs
/**
 * Rule filter evaluation — port of backend/core/rules.py
 * Evaluates forwarding rule filters against an ingest message.
 */

export function ruleMatchesMessage(filterJson, message) {
  const f = filterJson || {};

  const ingestIds = f.ingest_endpoint_ids;
  if (ingestIds != null) {
    const allowed = new Set(ingestIds.map(String));
    if (!allowed.has(String(message.ingest_endpoint_id || ""))) {
      return false;
    }
  }

  const bodyFilter = f.body || {};

  const contains = bodyFilter.contains;
  if (contains != null) {
    const hay = (message.body || "").toLowerCase();
    const needles = (Array.isArray(contains) ? contains : [])
      .map((s) => String(s).toLowerCase())
      .filter((s) => s.trim());
    if (needles.length > 0 && !needles.some((n) => hay.includes(n))) {
      return false;
    }
  }

  const regex = bodyFilter.regex;
  if (regex != null && String(regex).trim()) {
    try {
      const pat = new RegExp(String(regex), "i");
      if (!pat.test(message.body || "")) {
        return false;
      }
    } catch {
      return false;
    }
  }

  const priorityFilter = f.priority || {};
  const pmin = priorityFilter.min;
  if (pmin != null) {
    const minVal = parseInt(String(pmin), 10);
    if (!isNaN(minVal) && (message.priority || 3) < minVal) {
      return false;
    }
  }
  const pmax = priorityFilter.max;
  if (pmax != null) {
    const maxVal = parseInt(String(pmax), 10);
    if (!isNaN(maxVal) && (message.priority || 3) > maxVal) {
      return false;
    }
  }

  const tagsFilter = f.tags;
  if (tagsFilter != null && Array.isArray(tagsFilter) && tagsFilter.length > 0) {
    const msgTags = new Set(
      (Array.isArray(message.tags) ? message.tags : []).map((t) =>
        String(t).toLowerCase(),
      ),
    );
    const filterTags = tagsFilter.map((t) => String(t).toLowerCase());
    if (!filterTags.some((t) => msgTags.has(t))) {
      return false;
    }
  }

  const groupFilter = f.group;
  if (groupFilter != null) {
    if ((message.group || "") !== String(groupFilter)) {
      return false;
    }
  }

  return true;
}

File: backend/core/rules.py
import re

# Framework-agnostic rule matching logic
# Accepts dict representations instead of Django models


def rule_matches_message(filter_json: dict, message: dict) -> bool:
    """Check if message matches rule filter."""
    f = filter_json or {}

    ingest_ids = f.get("ingest_endpoint_ids")
    if ingest_ids is not None:
        if str(message.get("ingest_endpoint_id")) not in {str(x) for x in ingest_ids}:
            return False

    body_filter = f.get("body") or {}
    contains = body_filter.get("contains")
    if contains is not None:
        hay = (message.get("body") or "").lower()
        needles = [str(s).lower() for s in contains if str(s).strip()]
        if needles and not any(n in hay for n in needles):
            return False

    regex = body_filter.get("regex")
    if regex is not None and str(regex).strip():
        try:
            pat = re.compile(str(regex), flags=re.IGNORECASE)
        except re.error:
            return False
        if pat.search(message.get("body") or "") is None:
            return False

    priority_filter = f.get("priority") or {}
    pmin = priority_filter.get("min")
    if pmin is not None:
        try:
            if message.get("priority", 3) < int(pmin):
                return False
        except (ValueError, TypeError):
            pass
    pmax = priority_filter.get("max")
    if pmax is not None:
        try:
            if message.get("priority", 3) > int(pmax):
                return False
        except (ValueError, TypeError):
            pass

    tags_filter = f.get("tags")
    if tags_filter is not None and isinstance(tags_filter, list) and tags_filter:
        msg_tags = set()
        if isinstance(message.get("tags_json"), list):
            msg_tags = {str(t).lower() for t in message.get("tags_json", [])}
        filter_tags = {str(t).lower() for t in tags_filter}
        if not msg_tags & filter_tags:
            return False

    group_filter = f.get("group")
    if group_filter is not None:
        if (message.get("group") or "") != str(group_filter):
            return False

    return True
